### PR TITLE
melange: fix odoc

### PR DESCRIPTION
**Pull request by tatchi**
## Based on https://github.com/ocaml/dune/pull/7566

It fixes the failing test in #7566

### CODE REVIEW

File: test/blackbox-tests/test-cases/melange/odoc.t [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  $ dune build @doc
+
+  $ cat _build/default/_doc/_mlds/foo/index.mld
+  {0 foo index}
+  {1 Library foo}
+  The entry point of this library is the module:
+  {!module-Foo}.
+
+Fails for Melange-only libraries
```
Comment (jchavarri) [inline test/blackbox-tests/test-cases/melange/odoc.t:27]: I think this comment needs to be updated.

File: src/dune_rules/odoc.ml [hunk 2]
```diff
@@ -314,6 +319,12 @@ let setup_library_odoc_rules cctx (local_lib : Lib.Local.t) =
   let ctx = Super_context.context sctx in
   let* requires = Compilation_context.requires_compile cctx in
   let info = Lib.Local.info local_lib in
+  let modes = Lib_info.modes info in
+  let mode =
```
Comment (anmonteiro) [inline src/dune_rules/odoc.ml:323]: this needs to be done the other way around:

- it's only `` `Melange `` if `modes.melange == true` and the OCaml modes are false.

File: src/dune_rules/odoc.ml [hunk 3]
```diff
@@ -314,6 +319,12 @@ let setup_library_odoc_rules cctx (local_lib : Lib.Local.t) =
...
+    match modes.melange with
+    | false -> `Ocaml
+    | true -> `Melange
```
Comment (jchavarri) [inline src/dune_rules/odoc.ml:326]: Dune code style guide recommends keeping scope of bindings [as small as possible](https://github.com/ocaml/dune/blame/418a4e7f17b7f2e491fb0ef775fc3a7277b267c7/doc/hacking.rst#L504-L505). 

https://github.com/ocaml/dune/blob/418a4e7f17b7f2e491fb0ef775fc3a7277b267c7/doc/hacking.rst#L504-L505


So these 2 bindings could be moved to line 337, inside the definition of `compiled` binding.

**Review by jchavarri — APPROVED** [review]
Looks good, thanks! 2 minor requests.

File: src/dune_rules/odoc.ml (hunk 2 above)
Comment (anmonteiro) [reply src/dune_rules/odoc.ml:323]: we have `Lib_mode.Map.Set.for_merlin`. Looks like we want to share the same logic.

Comment (tatchi) [reply src/dune_rules/odoc.ml]: I ended up just re-using `Lib_mode.Map.Set.for_merlin`. Is that fine ?

Comment (anmonteiro) [reply src/dune_rules/odoc.ml]: Yup! that's even better. I wonder if we should rename it now that it's not for merlin only.

**Review by anmonteiro — APPROVED** [review]
should be OK to rename in a follow-up PR in case you can't think of a better name right now.

File: test/blackbox-tests/test-cases/melange/odoc.t [hunk 4]
```diff
@@ -0,0 +1,42 @@
+Test docs generation with Melange
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/odoc.t:1]: Could you add a dependency on the odoc package in this test?

**Review by Alizter — APPROVED** [review]

File: test/blackbox-tests/test-cases/melange/odoc.t (hunk 4 above)
Comment (tatchi) [reply test/blackbox-tests/test-cases/melange/odoc.t:1]: Done, assuming I correctly understood what you were asking for

File: test/blackbox-tests/test-cases/melange/dune [hunk 5]
```diff
@@ -9,3 +9,7 @@
 (cram
  (deps %{bin:melc_stdlib_prefix})
  (applies_to merlin))
+
+(cram
+ (deps (package odoc))
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/melange/dune:14]: @rgrinberg The odoc tests also rely on the odoc binary, so perhaps they should also be changed too.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/melange/dune:14]: How do you know the tests only rely on the binary?

**Pull request merged**